component/block_cache/block_cache.go

```diff
@@ -522,11 +522,15 @@ func (bc *BlockCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, e
 
 		// Copy data from this block to user buffer
 		readOffset := uint64(options.Offset) - block.offset
-		bytesRead := copy(options.Data[dataRead:], block.data[readOffset:])
+		bytesRead := copy(options.Data[dataRead:], block.data[readOffset:(block.endIndex-block.offset)])
 
 		// Move offset forward in case we need to copy more data
 		options.Offset += int64(bytesRead)
 		dataRead += bytesRead
+
+		if (block.endIndex - block.offset) < uint64(len(block.data)) {
```

Comment:
This check shall be corrected to compare options.offset == handle.size. When we are copying partial data from a block this condition can be true and it will return back EOF wrongly.